gh-115999: Refactor `LOAD_GLOBAL` specializations to avoid reloading {globals, builtins} keys

mpage opened the pull request:
Each of the `LOAD_GLOBAL` specializations is implemented roughly as:

1. Load keys version.
2. Load cached keys version.
3. Deopt if (1) and (2) don't match.
4. Load keys.
5. Load cached index into keys.
6. Load object from (4) at offset from (5).

This is not thread-safe in free-threaded builds; the keys object may be replaced in between steps (3) and (4).

This change refactors the specializations to avoid reloading the keys object and instead pass the keys object from guards to be consumed by downstream uops.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

mpage committed:
Refactor `LOAD_GLOBAL` specializations to avoid reloading {globals, builtins} keys

Each of the `LOAD_GLOBAL` specializations is implemented roughly as:

1. Load keys version.
2. Load cached keys version.
3. Deopt if (1) and (2) don't match.
4. Load keys.
5. Load cached index into keys.
6. Load object from (4) at offset from (5).

This is not thread-safe in free-threaded builds; the keys object may be replaced
in between steps (3) and (4).

This change refactors the specializations to avoid reloading the keys object and
instead pass the keys object from guards to be consumed by downstream uops.

colesbury commented on Python/bytecodes.c:
```diff
@@ -4778,6 +4787,26 @@ dummy_func(
+        tier2 op(_LOAD_GLOBAL_MODULE, (index/1 -- res, null if (oparg & 1))) {
+            PyDictObject *dict = (PyDictObject *)GLOBALS();
+            PyDictUnicodeEntry *entries = DK_UNICODE_ENTRIES(dict->ma_keys);
+            PyObject *res_o = entries[index].me_value;
+            DEOPT_IF(res_o == NULL);
+            Py_INCREF(res_o);
+            res = PyStackRef_FromPyObjectSteal(res_o);
```
We'll need to also update this pattern to make it thread-safe, but that can happen in subsequent PRs.

mpage replied:
Yep, I'm trying to keep PRs small and isolated to help minimize reviewer load.

colesbury commented on Python/executor_cases.c.h:
```diff
@@ -1700,17 +1721,22 @@
-            PyDictUnicodeEntry *entries = DK_UNICODE_ENTRIES(dict->ma_keys);
+            PyDictUnicodeEntry *entries = DK_UNICODE_ENTRIES(globals_keys);
             PyObject *res_o = entries[index].me_value;
             if (res_o == NULL) {
```
Is the deopt behavior here correct? The interpreter deopt in `generated_cases.c.h` looks okay because we don't actually modify the stack, but here in `executor_cases.c.h` we've pushed the keys object to the stack.

mpage replied:
Oof, yeah, that doesn't seem right. I assumed that because the uop consumed the keys from the stack it would also update stack pointer, but it doesn't look like that is the case. @brandtbucher @markshannon @Fidget-Spinner  - we chatted a bit about passing values between uops at the last sync - is there another way to do this?

mpage replied:
The `SYNC_SP()` I added should take care of this.

mpage committed:
Sync stack pointer in ops that consume keys

Ensure we update the stack to reflect that we've popped the keys.
There should be nothing on the stack if we deopt.

mpage committed:
Merge branch 'main' into gh-115999-refactor-load-global

mpage committed:
Mark keys dead

colesbury commented on Python/optimizer_cases.c.h:
```diff
@@ -844,30 +844,44 @@
-        case _GUARD_BUILTINS_VERSION: {
+        case _GUARD_GLOBALS_VERSION_PUSH_KEYS: {
+            PyDictKeysObject *globals_keys;
+            globals_keys = sym_new_not_null(ctx);
```
Looks like this is a problematic assignment:

https://github.com/python/cpython/actions/runs/11225895144/job/31205566003?pr=124953

mpage committed:
Fix compiler warning in optimizer_cases.c.h

Maybe this will stop tickling the msvc compiler bug, too?

colesbury reviewed (approved):
The [x86_64-pc-windows-msvc/msvc (Release)](https://github.com/python/cpython/actions/runs/11228012398/job/31211363688?pr=124953#logs) JIT failure is not new.

colesbury committed:
Merge branch 'main' into gh-115999-refactor-load-global

markshannon commented on Python/optimizer_analysis.c:
```diff
@@ -174,14 +194,18 @@ remove_globals(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer,
                 if (interp->rare_events.builtin_dict >= _Py_MAX_ALLOWED_BUILTINS_MODIFICATIONS) {
                     continue;
                 }
+                if (!check_next_uop(buffer, buffer_size, pc,
```
We want the optimizer passes to be (as much as possible) simple, fast scans over the uop sequence.
So, I'd like to avoid this sort of non-local check if possible.

Generally we want each pass to be a linear scan which maintains a small set of knowledge, like `function_checked`, etc. above.
Each case should then either update that knowledge or perform a simple optimization based on that knowledge.

FYI, I plan to merge this pass into optimizer_bytecodes.c which is also a linear pass with similar design principles (at least it should be).

markshannon commented on Python/bytecodes.c:
```diff
@@ -4871,6 +4884,26 @@ dummy_func(
             DEOPT_IF(func->func_version != func_version);
         }
 
+        tier2 op(_LOAD_GLOBAL_MODULE, (index/1 -- res, null if (oparg & 1))) {
```
This is effectively pushes the global's keys then does `_LOAD_GLOBAL_MODULE_FROM_KEYS`.

Maybe add a tier2 op that only pushes the keys? It might make the optimizer simpler as well.

markshannon commented on Python/bytecodes.c:
```diff
@@ -4871,6 +4884,26 @@ dummy_func(
+            null = PyStackRef_NULL;
+         }
+
+        tier2 op(_LOAD_GLOBAL_BUILTINS, (index/1 -- res, null if (oparg & 1))) {
```
Likewise

markshannon reviewed (commented):
Here's the review I did yesterday, but didn't quite finish in time 🙂
